### PR TITLE
AB#1087 Add intel quote provider library

### PR DESCRIPTION
**Pull request by daniel-weisse**
### Proposed changes
- Add the intel quote provider library to the ert docker images
- Pin the versions of installed intel sgx libraries

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -15,16 +15,25 @@ ENV LANGUAGE en_US:en
...
+    apt install -y \
+    az-dcap-client \
+    libsgx-dcap-ql=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-ql-dev=$DCAP_VERSION-bionic1 \
+    libsgx-pce-logic=$DCAP_VERSION-bionic1 \
+    libsgx-qe3-logic=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-quote-verify=$DCAP_VERSION-bionic1 \
+    libsgx-enclave-common=$PSW_VERSION-bionic1 \
+    libsgx-urts=$PSW_VERSION-bionic1 \
+    libsgx-enclave-common-dev=$PSW_VERSION-bionic1 \
```
Comment (thomasten) [inline Dockerfile:30]: Previously, the -dev packages were in stage sgx-dev such that they didn't get into the deploy target. Any need that you changed this?

**Review by m1ghtym0 — APPROVED** [review]
Agree with Thomas's comment on the dev packages, otherwise LGTM!

File: Dockerfile (hunk 1 above)
Comment (daniel-weisse) [reply Dockerfile:30]: I took another look at the build process:
`libsgx-qe3-logic` and `libsgx-pce-logic` are required by libsgx-dcap-ql, and therefore need to be installed in the sgx step.
`libsgx-ae-qve` `libsgx-ae-pce` and `libsgx-ae-qe3` already get installed by another package from the sgx step, even without `--no-install-recommends`
`libsgx-enclave-common-dev` is not required by anything else and i have moved it back to the sgx-dev step

File: Dockerfile [hunk 2]
```diff
@@ -15,13 +15,27 @@ ENV LANGUAGE en_US:en
...
 
 FROM common AS sgx
+ARG PSW_VERSION=2.13.103.1
+ARG DCAP_VERSION=1.10.103.1
 RUN apt update && \
-    apt install -y --no-install-recommends libsgx-dcap-ql libsgx-dcap-ql-dev libsgx-dcap-quote-verify libsgx-enclave-common libsgx-urts az-dcap-client && \
+    apt install -y --no-install-recommends \
+    az-dcap-client \
+    libsgx-dcap-ql=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-ql-dev=$DCAP_VERSION-bionic1 \
```
Comment (thomasten) [inline Dockerfile:24]: libsgx-dcap-ql-dev was there before, but is it required or can we also move it to sgx-dev?

File: Dockerfile [hunk 3]
```diff
@@ -15,13 +15,27 @@ ENV LANGUAGE en_US:en
...
+    libsgx-pce-logic=$DCAP_VERSION-bionic1 \
+    libsgx-qe3-logic=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-quote-verify=$DCAP_VERSION-bionic1 \
+    libsgx-enclave-common=$PSW_VERSION-bionic1 \
+    libsgx-urts=$PSW_VERSION-bionic1 \
+    libsgx-enclave-common-dev=$PSW_VERSION-bionic1 \
```
Comment (thomasten) [inline Dockerfile:30]:
```suggestion
```

File: Dockerfile [hunk 4]
```diff
@@ -15,13 +15,27 @@ ENV LANGUAGE en_US:en
...
+    libsgx-dcap-default-qpl=$DCAP_VERSION-bionic1 && \
     apt clean && apt autoclean
+# rename the softlink created by libsgx-dcap-default-qpl to avoid issues with az-dcap-client
+RUN mv /usr/lib/x86_64-linux-gnu/libdcap_quoteprov.so.1 /usr/lib/x86_64-linux-gnu/libdcap_quoteprov.so.1.intel
 
 FROM sgx AS sgx-dev
 RUN apt update && \
-    apt install -y libsgx-enclave-common-dev libsgx-ae-qve libsgx-ae-pce libsgx-ae-qe3 libsgx-qe3-logic libsgx-pce-logic && \
+    apt install -y libsgx-enclave-common-dev=$PSW_VERSION-bionic1 && \
```
Comment (thomasten) [inline Dockerfile:38]: Does this work? In my understanding PSW_VERSION is out of scope (https://docs.docker.com/engine/reference/builder/#scope). If so, we should declare it here, but define its value before the first FROM (https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact).

Comment (daniel-weisse) [reply Dockerfile:38]: The build was successful when i tried it. But from the docs it looks like it should cause a problem. 
Just to be safe i will add the arg to the build step

File: Dockerfile (hunk 2 above)
Comment (daniel-weisse) [reply Dockerfile:24]: Required for quote generation

File: Dockerfile [hunk 5]
```diff
@@ -15,13 +15,26 @@ ENV LANGUAGE en_US:en
...
+    apt install -y --no-install-recommends \
+    az-dcap-client \
+    libsgx-dcap-ql=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-ql-dev=$DCAP_VERSION-bionic1 \
+    libsgx-pce-logic=$DCAP_VERSION-bionic1 \
+    libsgx-qe3-logic=$DCAP_VERSION-bionic1 \
+    libsgx-dcap-quote-verify=$DCAP_VERSION-bionic1 \
+    libsgx-enclave-common=$PSW_VERSION-bionic1 \
+    libsgx-urts=$PSW_VERSION-bionic1 \
     apt clean && apt autoclean
```
Comment (thomasten) [inline Dockerfile:30]: * this seems broken
* did you remove libsgx-dcap-default-qpl on purpose?

File: Dockerfile [hunk 6]
```diff
@@ -15,13 +15,26 @@ ENV LANGUAGE en_US:en
...
+# rename the softlink created by libsgx-dcap-default-qpl to avoid issues with az-dcap-client
+RUN mv /usr/lib/x86_64-linux-gnu/libdcap_quoteprov.so.1 /usr/lib/x86_64-linux-gnu/libdcap_quoteprov.so.1.intel
 
 FROM sgx AS sgx-dev
+ARG PSW_VERSION=2.13.103.1
```
Comment (thomasten) [inline Dockerfile:35]:
```suggestion
ARG PSW_VERSION
```
suggest declaring the args in the stages, but defining them once before the stages as demonstrated here https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

**Review by thomasten — COMMENTED** [review]
pls also test this again

File: Dockerfile (hunk 5 above)
Comment (daniel-weisse) [reply Dockerfile]: deleted on accident, thanks for catching this.
Updated and works now

**Review by thomasten — APPROVED** [review]

**Pull request merged**